sklearn/decomposition/_pca.py

```diff
@@ -498,8 +498,14 @@ def _fit(self, X):
         # Handle svd_solver
         self._fit_svd_solver = self.svd_solver
         if self._fit_svd_solver == "auto":
+            # Automatically select arpack for sparse inputs
+            # Reduce n_components by 1 when passing from auto to arpack
+            if issparse(X):
+                self._fit_svd_solver = "arpack"
+                if self.n_components is None:
+                    n_components -= 1
```

Comment:
It may not be that intuitive because we determine `self._fit_svd_solver` based on `n_components` in some cases. Unless we do something like this (or did I miss something)?

```python
self._fit_svd_solver = self.svd_solver
if self._fit_svd_solver == "auto" and issparse(X):
    self._fit_svd_solver = "arpack"

# The original code for handling n_components=None here
# ...

# The original code for handling svd_solver here
# ...
```

Reply:
Indeed we do. You barebone code is the right one. Thanks @Charlie-XIAO

Reply:
Thanks for your suggestion. I cast the higher level ``` self.svd_solver``` to eliminate this ugly line ```n_components -=1```. Hope it looks better now.

Reply:
Thanks for updating, but note that according to [this principle](https://scikit-learn.org/dev/glossary.html#term-attribute):

> A public attribute on an estimator instance that does not end in an underscore should be the stored, unmodified value of an `__init__` parameter of the same name.

Hence you cannot directly modify `self.svd_solver`, but still need to use the pattern in https://github.com/scikit-learn/scikit-learn/pull/28498#discussion_r1506255770.